File: programmes_utiles/python/NearestNeighbour.py
from help import pathCost
from sys import maxsize
from time import process_time

def NearestNeighbour(tsp,n,start):
    path=[]
    path.append(start)
    nextTown=start

    t1_start = process_time()
    while len(path)<n:
        minTown=minimum(tsp,path,n,nextTown,start)
        path.append(minTown)
        nextTown=minTown
    t1_stop = process_time()

    path.append(start)
    totalCost=0
    for i in range(n):
        totalCost+=tsp[(path[i],path[i+1])]

    return path,totalCost,(t1_stop-t1_start)


def minimum(tsp,path,n,start,startTown):
    minCost=maxsize
    minPos=start
    for i in range(n):
        if start==i or startTown==i or i in path:
            continue
        if minCost>tsp[(start,i)]:
            minCost=tsp[(start,i)]
            minPos=i
    return minPos

File: src/AlgosC/C/NearestNeighbour/NearestNeighbour.c
#include <limits.h>
#include <stdio.h>
#include "../../utils/help.c"

int find_closest(int** tsp,int* path,int n,int start,int startTown){
    int minPos=start,minCost=INT_MAX;
    for (int i = 0; i < n; ++i) {
        if( start==i || startTown==i || existsIn(path,n,i)==1){
            continue;
        }

        if(minCost> tsp[start][i]){
            minCost=tsp[start][i];
            minPos=i;
        }
    }
    return minPos;
}
void NearestNeighbour(int** tsp,int n,int start,int *path){
    path[0]=start;
    int nextTown=start,lengthPath=0;
    while (lengthPath<n-1){
        int minTown=find_closest(tsp,path,n,nextTown,start);
        lengthPath++;
        path[lengthPath]=minTown;
        nextTown=minTown;
    }
}
